File: eureka/ios/nativeLib/nativeLib/EventEmittersToReact.swift
//
//  EventEmittersToReact.swift
//  eureka
//
//  Created by work on 04/02/21.
//

import Foundation

public class EventEmittersToReact {
  private static var parent: ((String, Any) -> Void)? = nil
      
  public init(pParent: @escaping (String, Any) -> Void){
    EventEmittersToReact.parent = pParent
  }
  
  internal class func debugLog(_ loggerStruct: LoggerStruct) {
    EventEmittersToReact.parent?("DebugLog", ["DebugLog": loggerStruct.getJSONStr()])
  }
    
  internal class func uploadOnCloud(_ data: String) {
    EventEmittersToReact.parent?("UploadOnCloud", ["UploadOnCloud": data])
  }
  
  public static func emitConnectResult(pResponse:ConnectResponse) {
    switch (pResponse.getResultCode().type) {
        case "intermediate":
            break;
        case "success":
//                LifePlusReactModule.cancelConnectTimeoutTimer();
//                BleCache.clearCharacteristicList();
          DeviceService.setCurrentProc(pCurrentProc: BleProcEnum.NONE);
            break;
        default:
//          Global.setUserId(pUserId: 0);
//                LifePlusReactModule.cancelConnectTimeoutTimer();
//                BleCache.clearCharacteristicList();
          DeviceService.setCurrentProc(pCurrentProc: BleProcEnum.NONE);
            break;
    }

    if (DeviceService.currentProcState != BleProcStateEnum.NONE) {
      DeviceService.setCurrentProcState(BleProcEnum.CONNECT, pCurrentProcState: pResponse.getResultCode().nextState);
    }
    EventEmittersToReact.parent?("ScanResult", ["ScanResult": pResponse.getResponseStr()])
  }
  
  internal static func EmitInstantMeasureResult(pResponse: InstantMeasureResponse) {
//    if (BleServices.getCurrentProcState() != BleProcStateEnum.NONE) {
//      BleServices.setCurrentProcState(BleProcEnum.INSTANT_MEASURE, pCurrentProcState: pResponse.getResultCode().nextState);
//    }
    EventEmittersToReact.parent?("InstantMeasureResult", ["InstantMeasureResult": pResponse.getResponseStr()])
  }
  
  //percent status
  internal static func percentStatus(percentage: Int) {
    EventEmittersToReact.parent?("PercentStatus", String(format: "{\"result\":{\"processcompletes\":\"%d%%\"}}", percentage))
  }
  
  internal static func EmitAppSyncResult(pResponse: AppSyncResponse, firmwareRevision: String){
    if (DeviceService.currentProcState != BleProcStateEnum.NONE) {
      DeviceService.setCurrentProcState(BleProcEnum.APP_SYNC, pCurrentProcState: pResponse.getResultCode().nextState)
    }
    EventEmittersToReact.parent?("AppSyncResult", ["InstantMeasureResult": pResponse.getResponseStr(), "FirmwareRevision": firmwareRevision])
  }

  internal class func EmitCalibrationResult(pResponse: CalibrationResponse) {
    EventEmittersToReact.parent?("CalibrationResult", ["CalibrationResult": pResponse.getResponseStr()])
  }

  internal class func EmitAutoCalibrationResult(pResponse:AutoCalibrationResponse) {
    EventEmittersToReact.parent?("AutoCalibrationResult", ["AutoCalibrationResult": pResponse.getResponseStr()])
  }

  internal class func EmitStepCount(pResponse: StepCountResponse) {
    EventEmittersToReact.parent?("StepCountResult", ["StepCountResult": pResponse.getResponseStr()])
  }

  internal class func EmitMealData(pResponse: MealDataResponse) {
    EventEmittersToReact.parent?("MealDataResult", ["MealDataResult": pResponse.getResponseStr()])
  }

  internal class func EmitCommonResult(pResponse: CommonResponse) {
    EventEmittersToReact.parent?("ScanResult", ["CommonResult": pResponse.getResponseStr()])
  }
    
  internal class func EmitFwUpdate(pResponse: FirmwareUpdateResponse) {
    EventEmittersToReact.parent?("FwUpdate", ["FwUpdate": pResponse.getResponseStr()])
  }
    
  public class func emitTimerTick() {
    EventEmittersToReact.parent?("TimerTick", [:])
  }
    
  public class func emitStepGoal(pResponse: StepGoalResponse) {
    EventEmittersToReact.parent?("StepGoalResult", ["StepGoalResult": pResponse.getResponseStr()])
  }
}

File: eureka/android/library/src/main/java/com/lifeplus/EventEmittersToReact.java
package com.lifeplus;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.lifeplus.Pojo.Responses.AppSyncResponse;
import com.lifeplus.Pojo.Responses.CalibrationResponse;
import com.lifeplus.Pojo.Responses.CommonResponse;
import com.lifeplus.Pojo.Responses.ConnectResponse;
import com.lifeplus.Pojo.Responses.FirmwareUpdateResponse;
import com.lifeplus.Pojo.Responses.InstantMeasureResponse;
import com.lifeplus.Pojo.Responses.MealDataResponse;
import com.lifeplus.Pojo.Responses.StepCountResponse;
import com.lifeplus.Pojo.Responses.StepGoalResponse;

public class EventEmittersToReact {
    private ReactContext _reactContext = null;
    private static final EventEmittersToReact instance = new EventEmittersToReact();

    public static EventEmittersToReact getInstance() {
        return instance;
    }

    public void setReactContext(ReactContext reactContext) {
        this._reactContext = reactContext;
    }

    public boolean isReady() {
        return _reactContext != null && _reactContext.hasActiveCatalystInstance();
    }

    public void debugLog(String data) {
        if (_reactContext != null && _reactContext.hasActiveCatalystInstance()) {
            WritableMap params = Arguments.createMap();
            params.putString("DebugLog", data);
            _reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("DebugLog", params);
        }
    }

    public void uploadOnCloud(String data) {
        if (_reactContext != null) {
            WritableMap params = Arguments.createMap();
            params.putString("UploadOnCloud", data);
            _reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("UploadOnCloud", params);
        }
    }

    //measures
    public void measuresEventEmitter(String toJson) {
        if (_reactContext != null) {
            WritableMap params = Arguments.createMap();
            params.putString("Measures", toJson);
            _reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("Measures", params);
        }
    }

    //percent status
    public void percentStatus(String toJson) {
        if (_reactContext != null) {
            WritableMap params = Arguments.createMap();
            params.putString("PercentStatus", toJson);
            _reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("PercentStatus", params);
        }
    }

    // Raw data transfer events
    public void RawDataTransferPercent(String toJson) {
        if (_reactContext != null) {
            WritableMap params = Arguments.createMap();
            params.putString("RawDataPercentage", toJson);
            _reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("Percentage", params);
        }
    }

    //emits after 100%
    public void measuresAfterPerCompletes(String toJson) {
        if (_reactContext != null) {
            WritableMap params = Arguments.createMap();
            params.putString("InstantMeasures", toJson);
            _reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("InstantMeasures", params);
        }
    }

    public void EmitConnectResult(ConnectResponse pResponse) {
        switch (pResponse.getResultCode().getType()) {
            case "intermediate":
                break;
            case "success":
//                LifePlusReactModule.cancelConnectTimeoutTimer();
//                BleCache.clearCharacteristicList();
//                BleServices.setCurrentProc(BleProcEnum.NONE);
                break;
            default:
                //Global.setUserId(0);
//                LifePlusReactModule.cancelConnectTimeoutTimer();
//                BleCache.clearCharacteristicList();
//                BleServices.setCurrentProc(BleProcEnum.NONE);
                break;
        }

        if (_reactContext != null) {
            WritableMap params = Arguments.createMap();
            params.putString("ScanResult", pResponse.getResponseStr());
            _reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("ScanResult", params);
        }
    }

    public void EmitInstantMeasureResult(InstantMeasureResponse pResponse) {
        if (_reactContext != null) {
            WritableMap params = Arguments.createMap();
            params.putString("InstantMeasureResult", pResponse.getResponseStr());
            _reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("InstantMeasureResult", params);
        }
    }

    public void EmitAppSyncResult(AppSyncResponse pResponse, String firmwareRevision) {
        if (_reactContext != null) {
            WritableMap params = Arguments.createMap();
            params.putString("InstantMeasureResult", pResponse.getResponseStr());
            params.putString("FirmwareRevision", firmwareRevision);
            _reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("AppSyncResult", params);
        }
    }

    public void EmitCommonResult(CommonResponse pResponse) {
        if (_reactContext != null) {
            WritableMap params = Arguments.createMap();
            params.putString("CommonResult", pResponse.getResponseStr());
            _reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("ScanResult", params);
        }
    }

    public void EmitCalibrationResult(CalibrationResponse pResponse) {
        if (_reactContext != null) {
            WritableMap params = Arguments.createMap();
            params.putString("CalibrationResult", pResponse.getResponseStr());
            _reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("CalibrationResult", params);
        }
    }

    public void EmitStepCount(StepCountResponse pResponse) {
        if (_reactContext != null) {
            WritableMap params = Arguments.createMap();
            params.putString("StepCountResult", pResponse.getResponseStr());
            _reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("StepCountResult", params);
        }
    }

    public void emitStepGoal(StepGoalResponse pResponse) {
        if (_reactContext != null) {
            WritableMap params = Arguments.createMap();
            params.putString("StepGoalResult", pResponse.getResponseStr());
            _reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("StepGoalResult", params);
        }
    }

    public void EmitMealData(MealDataResponse pResponse) {
        if (_reactContext != null) {
            WritableMap params = Arguments.createMap();
            params.putString("MealDataResult", pResponse.getResponseStr());
            _reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("MealDataResult", params);
        }
    }

    public void emitFwUpdate(FirmwareUpdateResponse pResponse){
        if (_reactContext != null) {
            WritableMap params = Arguments.createMap();
            params.putString("FwUpdate", pResponse.getResponseStr());
            _reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("FwUpdate", params);
        }
    }

    public void emitTimerTick(){
        if (_reactContext != null) {
            _reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("TimerTick", null);
        }
    }

    public void emitIncompatibleDevice(){
        if (_reactContext != null) {
            _reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("IncompatibleDevice", null);
        }
    }
}
